### PR TITLE
Optimize `yii\data\ActiveDataProvider`

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #12752 

Added tests to the #13467.

### CODE REVIEW

File: tests/framework/data/ActiveDataProviderTest.php [hunk 1]
```diff
@@ -177,4 +178,42 @@ public function testPaginationBeforeModels()
...
+            $this->assertCount(0, $provider->getModels());
+        } catch (InvalidCallException $exception) {
+            $this->fail('An excessive models query was executed.');
+        }
+
+        $this->assertEquals(0, $pagination->getPageCount());
+    }
+}
+
+class UnqueryableQueryMock extends Query
```
Comment (samdark) [inline tests/framework/data/ActiveDataProviderTest.php]: Class should be in its own file.

File: tests/framework/data/ActiveDataProviderTest.php [hunk 2]
```diff
@@ -177,4 +178,42 @@ public function testPaginationBeforeModels()
         $provider->refresh();
         $this->assertEquals(2, count($provider->getModels()));
     }
+
+    public function testZeroModels()
```
Comment (samdark) [inline tests/framework/data/ActiveDataProviderTest.php]: I'd name it `testNoQueryWhenZeroModels` or something like that.

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/framework/data/ActiveDataProviderTest.php (hunk 2 above)
Comment (Kolyunya) [inline tests/framework/data/ActiveDataProviderTest.php]: What about `testDoesNotPerformQueryWhenHasNoModels`? This seems to be more like a real English sentence.

File: tests/framework/data/ActiveDataProviderTest.php (hunk 1 above)
Comment (Kolyunya) [inline tests/framework/data/ActiveDataProviderTest.php]: Is `use yiiunit\framework\db\UnqueryableQueryMock`

**Comment by Kolyunya** [comment]
@samdark fixed mentioned issues.

File: tests/framework/data/ActiveDataProviderTest.php (hunk 1 above)
Comment (dynasource) [inline tests/framework/data/ActiveDataProviderTest.php]: I dont have a problem with keeping it in the current file as it is unlikely to be reused in other tests. Additionally it should be replaced by an anonymous class in PHP7 in the (hopefully near) future

File: tests/framework/data/ActiveDataProviderTest.php (hunk 2 above)
Comment (dynasource) [inline tests/framework/data/ActiveDataProviderTest.php]: lets keep it close to the code. I would prefer something like: `testSkipQueryWhenZeroTotalCount()`

File: tests/framework/data/ActiveDataProviderTest.php (hunk 1 above)
Comment (Kolyunya) [inline tests/framework/data/ActiveDataProviderTest.php]: @dynasource we may want to test the same thing (that a query was not performed)  if some other tests.

**Review by samdark — APPROVED** [review]

**Review by dynasource — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thanks!

**Pull request merged**